assistant: replace empty tool results with placeholder text

sharon-wang opened the pull request:
### Summary

- addresses https://github.com/posit-dev/positron/issues/7770
- improves message processing to handle empty content more effectively
    - we should not filter out _all_ empty responses
    - we can continue to filter out messages that have no content at all and messages that only contain empty text parts
    - however, we should not filter out tool results with empty content altogether
    - now, when a tool result part has empty content, we include placeholder content in place of the empty content, to ensure that tool calls will have corresponding tool results
- updates our workspace tools to return non-empty messages
    - throw an error when the tool criteria is not met
    - added text to tool descriptions to note that empty workspaces are not supported
- updates anthropic integration test to cover these cases
- adds whether workspace folders are open to the general context, so that LLMs are less likely to use the workspace tools when the workspace criteria is not met

### Release Notes

#### Bug Fixes

- Assistant: ensure tool calls have corresponding tool results (#7770)

### QA Notes

@:assistant

The easiest test case is running the project tree tool, file contents tool or text search tool when in a workspace-less window. The Assistant should not call any of these tools since no workspace folders are open. If you are able to sidestep this and get the Assistant to call the tools anyways, the tool should now throw a more user-friendly error, directing the user to ensure that a workspace folder is open.

sharon-wang committed:
assistant: replace empty message content with a placeholder

instead of filtering out empty content, replace it with a placeholder

sharon-wang committed:
specify message content for workspace tools

instead of returning empty content, include a helpful message describing why the tool is not returning the expected output

sharon-wang committed:
update comment

sharon-wang committed:
filter out empty text messages and use placeholder for empty tool results

instead of filtering out all empty messages, we will use a placeholder for empty tool results so that the LLM still receives a response corresponding to the tool call

sharon-wang committed:
update workspace tool descriptions to note empty workspace case

sharon-wang committed:
test fixes

sharon-wang commented on extensions/positron-assistant/src/test/anthropic.test.ts:
```diff
@@ -58,61 +59,24 @@ suite('AnthropicLanguageModel', () => {
 		sinon.restore();
 	});
 
-	test('provideLanguageModelResponse should filter out messages with empty text content', async () => {
```
I removed this test case because it is covered in the `provideLanguageModelResponse filters empty messages and different LanguageModelTextPart contents correctly` test

sharon-wang commented on src/vs/workbench/contrib/chat/browser/tools/fileContentsTool.ts:
```diff
@@ -59,7 +60,13 @@ export class FileContentsTool implements IToolImpl {
 			uri = URI.file(filePath);
 			if (!this.fileIsOpenOrInsideWorkspace(uri)) {
-				throw new Error(`File contents for ${filePath} can't be retrieved because the file is not open or inside the current workspace`);
+				return {
```
Returning a message felt better in the chat flow, as the Assistant will then suggest to the user to open up a workspace. When an error is returned, the Assistant doesn't say anything.

seeM replied:
I was initially unsure about when to throw vs early return in tools. The main difference is that throwing immediately halts the conversation, and returning gives the model a chance to reply.

Given the above, I think we should throw (with a helpful message!) when we're confident the user needs to take some action before trying again. It'll save them time and tokens! Also worth keeping in mind that models can sometimes go off on endless tool calling loops.

What do you think? An alternative here could be to keep the throw but also suggest how to remedy the error.

sharon-wang replied:
It makes sense to me to keep the throws, but add a bit more guidance on how to proceed. Coupled with the change to indicate in the context if workspace folders are open, I think we are far less likely to hit these errors in the workspace tools

seeM commented on src/vs/workbench/contrib/chat/browser/tools/projectTreeTool.ts:
```diff
@@ -23,6 +23,7 @@ This tool ignores node_modules, __pycache__, dist directories, certain files lik
 This tool does not provide information for specific files or directories, but rather gives an overview of the entire project structure.
 This tool is helpful for locating files and directories in the workspace.
 This tool only needs to be called once per conversation, unless files or directories are added, removed, moved, or renamed in the workspace.
+If the workspace is empty or no workspace folders are open, this tool cannot generate a project tree.
```
Another approach is to include workspace information in the context message and instruct the model not to call the tool if the context shows that there is no workspace open. I was able to get Anthropic to do less inspect variables requests for non-existent variables in that way:

> ...
> 
> If the user references a variable by name, first determine the `access_key` from the user context or a previous inspect variables result.
> 
> Do not call when:
>   - the variables do not appear in the user context
> ...

seeM replied:
Also, my intuition is that more direct instructions work better (and use fewer tokens) i.e.

> Do not use this tool when the workspace is empty or no workspace is open.

sharon-wang replied:
Ah this is much nicer! I rebased to pull in the context changes and was able to add a note to the context when workspace folders are not open. Now, with a corresponding update to the workspace tools descriptions, anthropic doesn't call any of the workspace tools if we've noted in the context that workspace folders aren't open 👌

seeM commented on extensions/positron-assistant/src/utils.ts:
```diff
@@ -256,22 +256,66 @@ export function isChatImageMimeType(mimeType: string): mimeType is vscode.ChatIm
 	return Object.values(vscode.ChatImageMimeType).includes(mimeType as vscode.ChatImageMimeType);
 }
 
+export const EMPTY_TOOL_RESULT_PLACEHOLDER = 'tool result is empty';
```
I wonder if an empty string is better, if it still fixes the issue. It'll use fewer tokens and IMO convey the same information.

sharon-wang committed:
add workspace folders check to context message

- to reduce tool call attempts to workspace tools that rely on a workspace folder being open, check and set workspace context when no workspace folders are open.

- update the workspace tools descriptions to be more direct with not using the tools when no workspace folders are open.

- throw an error when the workspace tools are used without a workspace folder open or without the file being available in the current workspace (instead of returning a message)

sharon-wang committed:
add more guidance to workspace tool error messages

sharon-wang committed:
add workspace context about workspace folders either way

sharon-wang committed:
use empty string as placeholder for empty tool result text

sharon-wang commented:
Thanks @seeM for your feedback! I have implemented all the suggestions, namely:

- add whether workspace folders are open into the context
- directly include in tool description `Do not use this tool when no workspace folders are open`
- throw errors in the tools when the tool criteria is not met (instead of returning messages)
- use an empty string as the empty tool result placeholder

Will merge once tests are green.

sharon-wang committed:
update the participant context tests to account for default content in the context
